opt, sql: support `UPDATE ... FROM` statements

ridwanmsharif opened the pull request:
Addresses #7841.

This change adds support for `UPDATE ... FROM` statements.
The FROM clause tables are joined together with the target
table and is used as input for the update. Furthermore, the
RETURNING clause can reference any table in the FROM clause.

Release note: None

Commit:
opt, sql: support `UPDATE ... FROM` statements

Addresses #7841.

This change adds support for `UPDATE ... FROM` statements.
The FROM clause tables are joined together with the target
table and is used as input for the update. Furthermore, the
RETURNING clause can reference any table in the FROM clause.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39202)
<!-- Reviewable:end -->

jordanlewis commented:
Awesome!

rytaft reviewed (commented):
Nice work!

> Reviewed 21 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlCYhwS1iZ9wcrVZnNq:bdtx8r6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, and @ridwanmsharif)

___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSFJIAGPW46RIsiXM:-LlCSFJIAGPW46RIsiXN:b16slrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L1)):*
> ```
> # LogicTest: local-opt
> ```

what about also adding `fakedist-opt`?
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSugq9O0nIiBdOCso:-LlCSugq9O0nIiBdOCsp:b2g0am4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L24)):*
> ```
> 
> statement ok
> INSERT INTO new_abc VALUES (1, 2, 3), (2, 3, 4)
> ```

[nit] might help to use different values for b and c so it's easier to distinguish from the original table
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSnPB9BtPhi_HtwHQ:-LlCSnPB9BtPhi_HtwHR:b-rkc5bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L40)):*
> ```
> 
> statement ok
> UPDATE abc SET b = other.b, c = other.c FROM new_abc AS other WHERE abc.a = other.a
> ```

what's the correct behavior in this case? It just uses the first matching row?
___
*[pkg/sql/opt/exec/execbuilder/mutation.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCUsLl9S18QWFSI5fz:-LlCUsLl9S18QWFSI5g-:b-djvza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/exec/execbuilder/mutation.go#L109)):*
> ```Go
> 	// TODO(andyk): Using ensureColumns here can result in an extra Render.
> 	// Upgrade execution engine to not require this.
> 	colList := make(opt.ColList, 0, len(upd.FetchCols)+len(upd.UpdateCols)+len(upd.CheckCols))
> ```

should you also add capacity for PassthroughCols?
___
*[pkg/sql/opt/ops/mutation.opt, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCVDOZ2LzwT4w9t-pg:-LlCVDOZ2LzwT4w9t-ph:b-rr6zgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/ops/mutation.opt#L125)):*
> ```
> 
>     # PassthroughCols are columns that the mutation needs to passthrough from
>     # its input. Its similar to the passthrough columns in projections. This
> ```

[nit] Its similar -> It's similar
___
*[pkg/sql/opt/optbuilder/delete.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCV_WfE8uLfo0euCCK:-LlCV_WfE8uLfo0euCCL:b-73kn8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/delete.go#L93)):*
> ```Go
> 	mb.outScope.expr = mb.b.factory.ConstructDelete(mb.outScope.expr, mb.checks, private)
> 
> 	mb.buildReturning(returning, nil)
> ```

nil -> nil /* fromCols */
___
*[pkg/sql/opt/optbuilder/insert.go, line 623 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCVla53_y33StAxCAz:-LlCVla53_y33StAxCB-:b-73kn8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/insert.go#L623)):*
> ```Go
> 	mb.outScope.expr = mb.b.factory.ConstructInsert(mb.outScope.expr, mb.checks, private)
> 
> 	mb.buildReturning(returning, nil)
> ```

nil -> nil /* fromCols */
___
*[pkg/sql/opt/optbuilder/insert.go, line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCVmYNCtEAtCI5NXqS:-LlCVmYNCtEAtCI5NXqT:b-73kn8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/insert.go#L874)):*
> ```Go
> 	mb.outScope.expr = mb.b.factory.ConstructUpsert(mb.outScope.expr, mb.checks, private)
> 
> 	mb.buildReturning(returning, nil)
> ```

nil -> nil /* fromCols */
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCW8-6-r0oQM-pskQZ:-LlCW8-6-r0oQM-pskQ_:bu23ef7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/mutation_builder.go#L192)):*
> ```Go
> // clause are returned.
> // TODO(andyk): Do needed column analysis to project fewer columns if possible.
> func (mb *mutationBuilder) buildInputForUpdate(
> ```

Can you extract some of the duplicated code into helper functions?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCWEMd0JsFO_rC6NYt:-LlCWEMd0JsFO_rC6NYu:b-h7k4qn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/mutation_builder.go#L294)):*
> ```Go
> 	// possible to remap columns, as in this example:
> 	//
> 	//   UPDATE abc SET a=b
> ```

this example should be changed to DELETE
___
*[pkg/sql/opt/xform/optimizer.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCWvy9BwItvvVDDYaD:-LlCWvy9BwItvvVDDYaE:b-3dophz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/xform/optimizer.go#L217)):*
> ```Go
> 
> 	// Validate there are no dangling references.
> 	if !root.Relational().OuterCols.Empty() && false {
> ```

I assume this was for debugging?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
Very nice! Thanks for taking care of this so quickly! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> overall, just some minor comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlD8WB98ytjPbuP8dlK:bww9rfg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj, @ridwanmsharif, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSugq9O0nIiBdOCso:-LlD7xw44PiwGvjm9tJm:b1l0rqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L24)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] might help to use different values for b and c so it's easier to distinguish from the original table
</blockquote></details>

Also, we should have some rows in `abc` that don't match and aren't updated.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSnPB9BtPhi_HtwHQ:-LlD72sD1uVzPzCp7TGL:b5s4a1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L40)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what's the correct behavior in this case? It just uses the first matching row?
</blockquote></details>

I'd add "an arbitrary row from each group is chosen (this is consistent with Postgres)" to the comment above
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD7JBB1UlziNcD9Bml:-LlD7JBB1UlziNcD9Bmm:b6z9fan) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L2)):*
> ```
> # LogicTest: local-opt
> ```

Could you also add some corresponding `execbuilder` tests that show the `EXPLAIN` for the statements in this test?
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD7oWa-xVdIgiZ0Cl5:-LlD7oWa-xVdIgiZ0Cl6:bz0r3rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L27)):*
> ```
> 
> statement ok
> UPDATE abc SET b = other.b, c = other.c FROM new_abc AS other WHERE abc.a = other.a
> ```

`AS other` is not necessary, and makes it harder to read
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD7-jXCjObFi598mCd:-LlD7-jXCjObFi598mCe:b-y4w22j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L50)):*
> ```
> # Returning old values.
> query IIIII colnames
> UPDATE abc SET b = old.b + 1, c = old.c + 2 FROM abc AS old WHERE abc.a = old.a RETURNING abc.a, abc.b AS new_b, old.b as old_b, abc.c as new_c, old.c as old_c
> ```

[nit] break this up into multiple lines
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD8NeCEOAcv7f_E5HN:-LlD8NeCEOAcv7f_E5HO:brugl1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L57)):*
> ```
> 
> # Check if RETURNING * returns everything
> query IIIIII
> ```

[nit] add `colnames` so the output is more readable
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD8Hfy7DeFlGiKdZwZ:-LlD8Hfy7DeFlGiKdZw_:b-ch0yz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L61)):*
> ```
> ----
> 1  4  7  1  3  5
> 2  5  8  2  4  6
> ```

I'd add a test or two with multiple tables in the FROM list
___
*[pkg/sql/opt/exec/factory.go, line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD4yEg-2UU9jT8Eo07:-LlD4yEg-2UU9jT8Eo08:b-13pht3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/exec/factory.go#L323)):*
> ```Go
> 	// columns in the same order as they appear in the table schema, with the
> 	// fetch columns first and the update columns second. The rowsNeeded parameter
> 	// is true if a RETURNING clause needs the updated row(s) as output.
> ```

Add an explanation of `passthrough`
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 201 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD5m6D4sOn8LNEj7SY:-LlD5m6D4sOn8LNEj7SZ:b-r1r3x0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/mutation_builder.go#L201)):*
> ```Go
> 	//
> 
> 	// FROM
> ```

this comment is confusing. I think it made sense for DELETE but not UPDATE
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 753 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD7T6sAvTImlEwuCqf:-LlD7T6sAvTImlEwuCqg:b-99vzs6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/mutation_builder.go#L753)):*
> ```Go
> // buildReturning wraps the input expression with a Project operator that
> // projects the given RETURNING expressions.
> func (mb *mutationBuilder) buildReturning(returning tree.ReturningExprs, fromCols []scopeColumn) {
> ```

[nit] document `fromCols`. Wondering if it makes more sense to name them `extraAccessibleCols` here
___
*[pkg/sql/opt/optbuilder/testdata/update_from, line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD6kurCwGSYkVncQeN:-LlD6kurCwGSYkVncQeO:b-amjefc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/testdata/update_from#L4)):*
> ```
> CREATE TABLE abc (a int primary key, b int, c int)
> ----
> ```

We need to test that the expressions inside the FROM list can't refer to columns from the table.
___
*[pkg/sql/opt/optbuilder/testdata/update_from, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD6R614yALU3nFyFci:-LlD6R62C-k2pRnEdWwS:b-y4w22j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/testdata/update_from#L77)):*
> ```
> # Check if UPDATE FROM works well with RETURNING expressions that reference the FROM tables.
> opt
> UPDATE abc SET b = old.b + 1, c = old.c + 2 FROM abc AS old WHERE abc.a = old.a RETURNING abc.a, abc.b AS new_b, old.b as old_b, abc.c as new_c, old.c as old_c
> ```

[nit] break this up into multiple lines


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
This is awesome, great job!!

high-level nit: I think it would be good if the change to support passthrough cols was a separate commit from the change to add UPDATE FROM, since they're conceptually pretty different, even though one depends on the other (if this is too much trouble, don't worry about it)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlCNQcD7K48h6Y_yf6l:b-xpywhr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj, @RaduBerinde, @ridwanmsharif, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSugq9O0nIiBdOCso:-LlDoWlABGO149o9taxd:b-hihe4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L24)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Also, we should have some rows in `abc` that don't match and aren't updated.
</blockquote></details>

I'm a big fan of making the different columns have different orders of magnitude, so it's easy to tell them apart (ex: `(1, 20, 300)`)
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDy-M270Kf6NLqoyu9:-LlDy-M331cBjKC4b95J:b-jmgir6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L40)):*
> ```
> 
> statement ok
> UPDATE abc SET b = other.b, c = other.c FROM new_abc AS other WHERE abc.a = other.a
> ```

is `LATERAL` valid here? If yes, should it be tested?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlE1DmJ4r3lr3eqvGEF:-LlE1DmJ4r3lr3eqvGEG:bz8nr71) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/mutation_builder.go#L189)):*
> ```Go
> //
> // All columns from the table to update are added to fetchColList.
> // If a FROM clause is defined, the columns in the FROM
> ```

nit: this comment could be a little more descriptive, like where the columns come from, how the data source is constructed, etc.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDth9O0d0kesVHljAO:-LlDth9O0d0kesVHljAP:b-8b58eo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/mutation_builder.go#L213)):*
> ```Go
> 	numCols := len(mb.outScope.cols)
> 
> 	// If there is a from clause present, we must join all the tables together with the
> ```

nit: `FROM`
___
*[pkg/sql/opt/optbuilder/update.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDvLu-3HTmH_7zXxqW:-LlDvLu07BcWGHzPuQHt:b-yk703i) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/update.go#L106)):*
> ```Go
> 	//
> 	// All columns from the update table will be projected.
> 	fromCols := mb.buildInputForUpdate(inScope, upd.From, upd.Where, upd.Limit, upd.OrderBy)
> ```

can this be accessed via `mb.outScope.cols`?
___
*[pkg/sql/opt/optbuilder/testdata/update_from, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD6R614yALU3nFyFci:-LlDwEJlDixrVbbL0eLz:b-rils2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/testdata/update_from#L77)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] break this up into multiple lines
</blockquote></details>

I generally use vim to edit these test files, and I have this in my vimrc to sqlfmt highlighted queries:
```
noremap ,s :!cockroach sqlfmt --print-width 80 --use-spaces<cr><cr>
```
___
*[pkg/sql/parser/sql.y, line 5526 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDw_v5-MFI4dhsu-01:-LlDw_v60avyWmyMvH6L:bj418xw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/parser/sql.y#L5526)):*
> ```Yacc
> update_stmt:
>   opt_with_clause UPDATE table_name_expr_opt_alias_idx
>     SET set_clause_list opt_from_list opt_where_clause opt_sort_clause opt_limit_clause returning_clause
> ```

did you confirm with PG that this is the right place for this? i've gotten these ordered wrong in the past
___
*[pkg/sql/sem/tree/pretty.go, line 1027 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDx2DjElleCe6DXb4U:-LlDx2DjElleCe6DXb4V:b2ezo0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/sem/tree/pretty.go#L1027)):*
> ```Go
> 		p.row("UPDATE", p.Doc(node.Table)),
> 		p.row("SET", p.Doc(&node.Exprs)))
> 	if len(node.From) > 0 {
> ```

Are there tests for this? see `sem/tree/testdata/pretty`


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlE_WYN-z0axa7ZXi34:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSFJIAGPW46RIsiXM:-LlDiKXI6hFHgffCPLb-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what about also adding `fakedist-opt`?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSugq9O0nIiBdOCso:-LlEYrRj0ZjBXnDd67o4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L24)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I'm a big fan of making the different columns have different orders of magnitude, so it's easy to tell them apart (ex: `(1, 20, 300)`)
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCSnPB9BtPhi_HtwHQ:-LlDkez6DGvsllZhsAvz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/logictest/testdata/logic_test/update_from#L40)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'd add "an arbitrary row from each group is chosen (this is consistent with Postgres)" to the comment above
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD7oWa-xVdIgiZ0Cl5:-LlDkgfB14uh3HPqZnGV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L27)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`AS other` is not necessary, and makes it harder to read
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDy-M270Kf6NLqoyu9:-LlEXD4M7Hp_trd9Nd-y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L40)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

is `LATERAL` valid here? If yes, should it be tested?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD7-jXCjObFi598mCd:-LlDo1uK8TnhYn77uia_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L50)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] break this up into multiple lines
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD8NeCEOAcv7f_E5HN:-LlDo2r72wCc7VnP48Mw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L57)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add `colnames` so the output is more readable
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD8Hfy7DeFlGiKdZwZ:-LlDszCI7jHtcNoWeRN1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L61)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'd add a test or two with multiple tables in the FROM list
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/factory.go, line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD4yEg-2UU9jT8Eo07:-LlDtWZEBV5TmRVD4MHs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/exec/factory.go#L323)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add an explanation of `passthrough`
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/mutation.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCUsLl9S18QWFSI5fz:-LlDtifs7j2hObZ0n_tf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/exec/execbuilder/mutation.go#L109)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should you also add capacity for PassthroughCols?
</blockquote></details>

Done.
___
*[pkg/sql/opt/ops/mutation.opt, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCVDOZ2LzwT4w9t-pg:-LlDtnf00A5nxuo-aM-z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/ops/mutation.opt#L125)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] Its similar -> It's similar
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/delete.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCV_WfE8uLfo0euCCK:-LlDvN9zBeTKS_RAXFmt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/delete.go#L93)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nil -> nil /* fromCols */
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/insert.go, line 623 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCVla53_y33StAxCAz:-LlDvNndCxVz8csrEGbx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/insert.go#L623)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nil -> nil /* fromCols */
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/insert.go, line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCVmYNCtEAtCI5NXqS:-LlDvO8mAUmYfg4scffO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/insert.go#L874)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nil -> nil /* fromCols */
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCW8-6-r0oQM-pskQZ:-LlDvmL48ti_kYiryZlX:bl8vkw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/mutation_builder.go#L192)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you extract some of the duplicated code into helper functions?
</blockquote></details>

It becomes a little ugly when I do that. Like the only part that's common is the `buildWhere` part and the `SELECT + ORDER` part. `buildInputForUpdate` then applies a distinct on on top of that where as `buildInputForDelete` doesn't.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCWEMd0JsFO_rC6NYt:-LlDwJdH74KAZEsHLt27:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/mutation_builder.go#L294)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this example should be changed to DELETE
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlE1DmJ4r3lr3eqvGEF:-LlE_DYe1uA95HxHjOky:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/mutation_builder.go#L189)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: this comment could be a little more descriptive, like where the columns come from, how the data source is constructed, etc.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDth9O0d0kesVHljAO:-LlEXqax96QxFIPA7jKV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/mutation_builder.go#L213)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: `FROM`
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 753 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD7T6sAvTImlEwuCqf:-LlDwLXr1mRnWe8dUgHJ:b38jhzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/mutation_builder.go#L753)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] document `fromCols`. Wondering if it makes more sense to name them `extraAccessibleCols` here
</blockquote></details>

That's good. Done. Naming is hard.
___
*[pkg/sql/opt/optbuilder/update.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDvLu-3HTmH_7zXxqW:-LlEXteH1Rfxb3WtWr5p:b-lyqdsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/update.go#L106)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

can this be accessed via `mb.outScope.cols`?
</blockquote></details>

Well yes but we want just the `FROM` columns to be extracted and so we do it before they are merged with `mb.outScope.cols`.
___
*[pkg/sql/opt/optbuilder/testdata/update_from, line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD6kurCwGSYkVncQeN:-LlDzp_FCipx1jT34ADg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/testdata/update_from#L4)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We need to test that the expressions inside the FROM list can't refer to columns from the table.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/update_from, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD6R614yALU3nFyFci:-LlEYHII2my1yYGuaJ--:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/opt/optbuilder/testdata/update_from#L77)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I generally use vim to edit these test files, and I have this in my vimrc to sqlfmt highlighted queries:
```
noremap ,s :!cockroach sqlfmt --print-width 80 --use-spaces<cr><cr>
```
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/optimizer.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCWvy9BwItvvVDDYaD:-LlDwXd25cnAjmeauBQF:b-gityk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/xform/optimizer.go#L217)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I assume this was for debugging?
</blockquote></details>

omg yes thank you. I should've caught that. I was subconsciously relying on some linter to remind me to get rid of that.
___
*[pkg/sql/parser/sql.y, line 5526 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDw_v5-MFI4dhsu-01:-LlE2TDt3hGJZm904SKv:b4j6h18) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/parser/sql.y#L5526)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

did you confirm with PG that this is the right place for this? i've gotten these ordered wrong in the past
</blockquote></details>

Yep. Right place.
___
*[pkg/sql/sem/tree/pretty.go, line 1027 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlDx2DjElleCe6DXb4U:-LlEQDnYCp8iGpoKtC4x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/sem/tree/pretty.go#L1027)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Are there tests for this? see `sem/tree/testdata/pretty`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlEp_dT7MkPfan1jkPY:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj, @ridwanmsharif, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlD8NeCEOAcv7f_E5HN:-LlEotv-88XTBZoFEXqW:bq6plua) ([raw file](https://github.com/cockroachdb/cockroach/blob/4118f6100bc0c480de9b9427ec0c76af300bdc9b/pkg/sql/logictest/testdata/logic_test/update_from#L57)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think it needs to be `colnames,rowsort`
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlEpFh50x214mGltBLX:-LlEpFh679jSQQlLjo-a:bl1d67l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bee9a2cb19aa04f4c222fc4d521836b94948b4f4/pkg/sql/logictest/testdata/logic_test/update_from#L87)):*
> ```
> 
> # Check if RETURNING * returns everything
> query IIIIII colnames rowsort
> ```

`colnames,rowsort`
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlEpKSH6f8A4t2O8jrM:-LlEpKSH6f8A4t2O8jrN:bl1d67l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bee9a2cb19aa04f4c222fc4d521836b94948b4f4/pkg/sql/logictest/testdata/logic_test/update_from#L149)):*
> ```
> 
> # Make sure UPDATE ... FROM works with LATERAL.
> query IIIIIII colnames rowsort
> ```

`colnames,rowsort`
___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 168 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlEonhi4LvIg5_-e_21:-LlEonhi4LvIg5_-e_22:b-7qa5hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bee9a2cb19aa04f4c222fc4d521836b94948b4f4/pkg/sql/logictest/testdata/logic_test/update_from#L168)):*
> ```
> # Make sure the FROM clause cannot reference the target table.
> statement error no data source matches prefix: abc
> UPDATE abc SET a = other.a FROM (SELECT abc.a from abc as x) as other WHERE abc.a=other.a
> ```

[nit] AS


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
What are these new files that start with 22?

> Reviewed 1 of 1 files at r2, 19 of 20 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlHJbmM38shLk3PONut:b-addhvx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj and @ridwanmsharif)

___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHGXXU0RoB_3zWvOvK:-LlHGXXV3s4D0NzKmwjk:b-5n42ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/logictest/testdata/logic_test/update_from#L118)):*
> ```
> # Update values of table from values expression
> statement ok
> UPDATE abc SET b = other.b, c = other.c FROM (values (1, 2, 3), (2, 3, 4)) as other ("a", "b", "c") WHERE abc.a = other.a
> ```

[nit] what did `abc` contain before this statement was executed? Might be good to just print out the value of `SELECT * FROM abc` before this.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCW8-6-r0oQM-pskQZ:-LlHIZ6H91h1duUOsmM2:b2nhlkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/mutation_builder.go#L192)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

It becomes a little ugly when I do that. Like the only part that's common is the `buildWhere` part and the `SELECT + ORDER` part. `buildInputForUpdate` then applies a distinct on on top of that where as `buildInputForDelete` doesn't.
</blockquote></details>

Ok!
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlCWEMd0JsFO_rC6NYt:-LlHIf_j00kbDyLmLJTd:b4kahzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad4246e147ff6762aeb381c3869a8d977f40c17/pkg/sql/opt/optbuilder/mutation_builder.go#L294)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This isn't correct SQL anymore
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 200 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHIS6ACsljr_gnXEM0:-LlHIS6ACsljr_gnXEM1:b-bu20z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/optbuilder/mutation_builder.go#L200)):*
> ```Go
> // predictable, consistent with the POSTGRES implementation).
> // buildInputForUpdate returns the columns of the FROM tables so they
> // can me made accessible to other parts of the query (RETURNING clause).
> ```

me -> be
___
*[pkg/sql/opt/optbuilder/update.go, line 329 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHJ2QQ1CaEu2xWU5v0:-LlHJ2QRAZ3Py0J7S50i:b7n8dl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/optbuilder/update.go#L329)):*
> ```Go
> 	private := mb.makeMutationPrivate(returning != nil)
> 	for _, col := range fromCols {
> 		if col.id != 0 && !col.hidden {
> ```

Is it not possible for the returning clause to refer to a hidden column explicitly?


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj and @ridwanmsharif)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 202 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHrXD29uRISqZZk2l4:-LlHrXD29uRISqZZk2l5:b-tclg6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/optbuilder/mutation_builder.go#L202)):*
> ```Go
> // can me made accessible to other parts of the query (RETURNING clause).
> // TODO(andyk): Do needed column analysis to project fewer columns if possible.
> func (mb *mutationBuilder) buildInputForUpdate(
> ```

I think you should instead make a `mutationBuilder.fromCols` field, since that's how mutation builder works in every other case. Why should `fromCols` be passed around as params/return values, when `tab`, `tabID`,  `alias`, `outScope`, `fetchOrds`, etc. all be fields on `mutationBuilder`? I don't think it's any more special than any of those.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj and @ridwanmsharif)

___
*[pkg/sql/opt/ops/mutation.opt, line 129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHtPEH917W9yod3EO7:-LlHtPEH917W9yod3EO8:bgskenf) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/ops/mutation.opt#L129)):*
> ```
>     # references columns from tables in the `FROM` clause. When this happens
>     # the update will need to pass through those refenced columns from its input.
>     PassthroughCols ColList
> ```

This feels awkward to me to have both `ReturnCols` and `PassthroughCols`. Why isn't it just `ReturnCols`? These mutation operators are already complex and hard to reason about it, and segregating these two collections makes them even tougher.


<!-- Sent from Reviewable.io -->

Commit:
opt: returning clause of UPDATE ... FROM can reference hidden cols

This change adds the hidden columns to the outscope of a distinct
on and allows UPDATE ... FROM queries to reference hidden columns
in the RETURNING clause.

Release note: None

Commit:
opt: prune passthrough columns of a mutation

This change prunes the passthrough columns of a mutation
when it prunes away the returning columns.

Release note: None

ridwanmsharif reviewed (commented):
TFTRs! Also the files added were the test files for the pretty formatting of `UPDATE...FROM` queries.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlUhQ9_455bAV7pyldv:b-8if790)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/update_from, line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHGXXU0RoB_3zWvOvK:-LlUIgwi-B6MlqYbgJom:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/logictest/testdata/logic_test/update_from#L118)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] what did `abc` contain before this statement was executed? Might be good to just print out the value of `SELECT * FROM abc` before this.
</blockquote></details>

Done.
___
*[pkg/sql/opt/ops/mutation.opt, line 129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHtPEH917W9yod3EO7:-LlUJuP2DDwKdXQl2m4p:bqxnh2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/ops/mutation.opt#L129)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This feels awkward to me to have both `ReturnCols` and `PassthroughCols`. Why isn't it just `ReturnCols`? These mutation operators are already complex and hard to reason about it, and segregating these two collections makes them even tougher.
</blockquote></details>

There are a few places where we require the `ReturnCols` to have exactly the number of columns as columns in the table (we have 0 values for columns not returned). Also we need to communicate with exec, the passthrough columns from the input we want to fetch separately. We could have the passthrough columns be part of the `ReturnCols` but it would make the them even more complex to reason about (the first |table columns| columns being part of the table and the rest from the passthrough set). Keeping them separate seems cleaner and is also similar to what we do with projections, and it also makes the pruning logic easier to understand.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 200 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHIS6ACsljr_gnXEM0:-LlUJOIy15YwdNtH05rT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/optbuilder/mutation_builder.go#L200)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

me -> be
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 202 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHrXD29uRISqZZk2l4:-LlUJPjO4afeIJDvWEGn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/optbuilder/mutation_builder.go#L202)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think you should instead make a `mutationBuilder.fromCols` field, since that's how mutation builder works in every other case. Why should `fromCols` be passed around as params/return values, when `tab`, `tabID`,  `alias`, `outScope`, `fetchOrds`, etc. all be fields on `mutationBuilder`? I don't think it's any more special than any of those.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/update.go, line 329 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlHJ2QQ1CaEu2xWU5v0:-LlUJo86EWoRhqcS5Rec:b-q51v21) ([raw file](https://github.com/cockroachdb/cockroach/blob/910e2711f038af7d3e5aface480093bdf63c9111/pkg/sql/opt/optbuilder/update.go#L329)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is it not possible for the returning clause to refer to a hidden column explicitly?
</blockquote></details>

It wasn't possible before because of the distinct on. I added a change to allow this.


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/optbuilder/distinct.go, line 115 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlUzhcZ4iwNjhgRLisN:-LlUzhcZ4iwNjhgRLisO:b-vrgshk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b9a95bd7eb2cfa6d544fe7217852a85ec3b76f4/pkg/sql/opt/optbuilder/distinct.go#L115)):*
> ```Go
> 	// Add the output columns.
> 	for i := range inScope.cols {
> 		if !inScope.cols[i].hidden {
> ```

I don't see why we excluded hidden columns before? Am I missing something here?


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work!

> Reviewed 7 of 9 files at r5, 2 of 4 files at r6, 5 of 5 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlWeimXD__e80JLG2ey:b-kbynj8)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/distinct.go, line 115 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlUzhcZ4iwNjhgRLisN:-LlWdj2mBFzVuNFAyXA3:brc2i7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b9a95bd7eb2cfa6d544fe7217852a85ec3b76f4/pkg/sql/opt/optbuilder/distinct.go#L115)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

I don't see why we excluded hidden columns before? Am I missing something here?
</blockquote></details>

I'm not sure -- I think @RaduBerinde wrote this code


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-:-LlXHlJsANw2iYHO1vxV:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/distinct.go, line 115 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/39202#-LlUzhcZ4iwNjhgRLisN:-LlXHoKq4lh1-_6ndItz:b-mlfx0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b9a95bd7eb2cfa6d544fe7217852a85ec3b76f4/pkg/sql/opt/optbuilder/distinct.go#L115)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I'm not sure -- I think @RaduBerinde wrote this code
</blockquote></details>

We spoke offline, it doesn't seem that there's a reason we couldn't include the hidden columns here.


<!-- Sent from Reviewable.io -->

knz commented:
Curious: is there going to be a followup for DELETE FROM, which uses the same planning rules?
